### PR TITLE
feat(rtcstats): report pc connection failure

### PR DESCRIPTION
**Pull request by andrei-gavrilescu**

### CODE REVIEW

File: react/features/rtcstats/RTCStats.ts [hunk 1]
```diff
@@ -16,6 +25,8 @@ import {
     VideoTypeData
 } from './types';
 
+declare const APP: any;
```
Comment (saghul) [inline react/features/rtcstats/RTCStats.ts:28]: we no longer need this

File: react/features/rtcstats/RTCStats.ts [hunk 2]
```diff
@@ -185,6 +198,38 @@ class RTCStats {
...
+    /**
+     * RTCStats client can notify the APP of any PeerConnection related event that occurs.
+     *
+     * @param {Object} event - Rtcstats event.
+     * @returns {void}
+     */
+    handleRtcstatsEvent(event: any) {
+        switch (event.type) {
+        case PC_CON_STATE_CHANGE: {
+            const { body: { isP2P, state } = { state: null,
```
Comment (saghul) [inline react/features/rtcstats/RTCStats.ts:210]: Do we get any other data that would be worthy of extraction as part of the event?

File: react/features/rtcstats/RTCStats.ts [hunk 3]
```diff
@@ -185,6 +198,38 @@ class RTCStats {
...
+            // it usually means that none of our services can be reached i.e. there's problem with the internet
+            // connection and not necessarily with reaching the JVB (due to a firewall or other reasons).
+            if (state === PC_STATE_FAILED && this.trace.isConnected()) {
+                const connectionType = isP2P ? 'P2P' : 'JVB';
+                const wasConnected = this.connStateEvents.some((connectionEvent: { isP2P: any; state: string; }) =>
+                    (connectionEvent.isP2P === isP2P) && (connectionEvent.state === PC_STATE_CONNECTED));
+
+                logger.info(`${connectionType} PeerConnection failed, previously connected: ${wasConnected}`);
+
+                APP.API.notifyPeerConnectionFailure(isP2P, wasConnected);
```
Comment (saghul) [inline react/features/rtcstats/RTCStats.ts:225]: You need to check if typeof APP is undefined before calling this or mobile will die.

**Review by saghul — CHANGES_REQUESTED** [review]

File: react/features/rtcstats/RTCStats.ts (hunk 2 above)
Comment (andrei-gavrilescu) [reply react/features/rtcstats/RTCStats.ts:208]: there's also a connection identifier (pcId) but didn't see any use for it for now.

File: react/features/rtcstats/RTCStats.ts (hunk 3 above)
Comment (andrei-gavrilescu) [reply react/features/rtcstats/RTCStats.ts]: ow, we don't want that, fixed.

File: react/features/rtcstats/RTCStats.ts (hunk 1 above)
Comment (andrei-gavrilescu) [reply react/features/rtcstats/RTCStats.ts]: removed

File: react/features/rtcstats/RTCStats.ts [hunk 4]
```diff
@@ -185,6 +196,38 @@ class RTCStats {
...
+            // it usually means that none of our services can be reached i.e. there's problem with the internet
+            // connection and not necessarily with reaching the JVB (due to a firewall or other reasons).
+            if (state === PC_STATE_FAILED && this.trace.isConnected()) {
+                const connectionType = isP2P ? 'P2P' : 'JVB';
+                const wasConnected = this.connStateEvents.some((connectionEvent: { isP2P: any; state: string; }) =>
+                    (connectionEvent.isP2P === isP2P) && (connectionEvent.state === PC_STATE_CONNECTED));
+
+                logger.info(`${connectionType} PeerConnection failed, previously connected: ${wasConnected}`);
+
+                APP?.API?.notifyPeerConnectionFailure(isP2P, wasConnected);
```
Comment (saghul) [inline react/features/rtcstats/RTCStats.ts:223]: This won't cut it, you'd need to optional chain the function call too. Better to check if APP is undefined or not.

Comment (andrei-gavrilescu) [reply react/features/rtcstats/RTCStats.ts:223]: I'm a bit confused, isn't the first part `APP?.API` the same as checking if APP is undefined or not?

Comment (saghul) [reply react/features/rtcstats/RTCStats.ts:223]: That works if there is a variable called APP which is undefined. But APP is not even declared.

Comment (andrei-gavrilescu) [reply react/features/rtcstats/RTCStats.ts]: understood, fixed.

**Review by saghul — APPROVED** [review]

**Pull request merged**